File: Leetcode/1319 Number of Operations to Make Network Connected.py
class Solution:
    def makeConnected(self, n, connections):
        if len(connections) < (n - 1):
            return -1

        adj_list = get_adj_list(n, connections)
        visited = [False] * n
        
        n_connected_components = 0
        for start in range(n):
            if not visited[start]:
                dfs(start, adj_list, visited)
                n_connected_components += 1
        
        return n_connected_components - 1


def get_adj_list(n, edges):
    adj_list = [[] for _ in range(n)]
    for src, dst in edges:
        adj_list[src].append(dst)
        adj_list[dst].append(src)
    return adj_list

def dfs(start, adj_list, visited):
    if visited[start]:
        return
    visited[start] = True
    for child in adj_list[start]:
        dfs(child, adj_list, visited)

File: Leetcode/1319 Number of Operations to Make Network Connected.cpp
class Solution {
public:
    int makeConnected(int n, const vector<vector<int>> &connections) {
        if (connections.size() < n - 1) {
            return -1;
        }
        auto adjList = getAdjList(n, connections);
        vector<bool> visited(n, false);
        int nConnectedComponents = 0;
        for(int start = 0; start < n; start++) {
            if(!visited[start]) {
                dfs(start, adjList, visited);
                nConnectedComponents++;
            }
        }
        return nConnectedComponents - 1;
    }

    vector<vector<int>> getAdjList(int n, const vector<vector<int>> &edges) {
        vector<vector<int>> adjList(n, vector<int>());
        for(auto &edge : edges) {
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
        }
        return adjList;
    }

    void dfs(int start, const vector<vector<int>> &adjList, vector<bool>& visited) {
        if (visited[start]) {
            return;
        }
        visited[start] = true;
        for(auto child : adjList[start]) {
            dfs(child, adjList, visited);
        }
    }
};
